### DISCUSSION TITLE
What does it mean when git lfs ls-files lists nothing but git lfs checkout reports checking out objects?

### DISCUSSION THREAD

**Discussion by gpakosz** [git-lfs/git-lfs#5301] [General] [answered]
Hello,

What does it mean when in a repository, `git lfs ls-files` lists nothing but `git lfs checkout` reports
```
$ git lfs checkout
Checking out LFS objects: 100% (15/15), 0 B | 0 B/s, done.
```
?

**Comment by chrisd8088** [comment, marked as answer]
Hey, thanks for the question.  I suspect this means that you have some empty (zero-byte) files whose filenames match the `.gitattributes` patterns for files to be tracked by Git LFS.  However, Git LFS will not actually create a pointer for those empty files, so nothing is reported by `git lfs ls-files`.

Here's a simple example which might represent what is going on for you:
```
$ git init test
$ cd test
$ git lfs track '*.bin'
$ touch foo.bin
$ git add foo.bin .gitattributes
$ git commit -m init

$ git lfs ls-files
$ git show HEAD:foo.bin
$ git lfs checkout
Checking out LFS objects: 100% (1/1), 0 B | 0 B/s, done.                        

$ echo bar >bar.bin
$ git add bar.bin
$ git commit -m bar

$ git lfs ls-files
7d865e959b * bar.bin
$ git show HEAD:bar.bin
version https://git-lfs.github.com/spec/v1
oid sha256:7d865e959b2466918c9863afca942d0fb89d7c9ac0c99bafc3749504ded97730
size 4
$ git lfs checkout
Checking out LFS objects: 100% (2/2), 4 B | 0 B/s, done.  
```

**Comment by gpakosz** [comment]
Hello @chrisd8088,

Thanks for your answer.
There are indeed `15` empty files as reported by `find . -type f -empty | wc -l`.

The `.gitattributes` files doesn't contain any entry specifying `filter=lfs` though.
Which is confirmed by the following command?
```
$ git ls-files ':(attr:filter=lfs)'
```

**Reply by chrisd8088** [reply]
Ah, that's a good point, I forgot that since v3.x Git LFS recognizes zero-byte files regardless of their tracked status.
```
$ git init test
$ cd test
$ touch foo.bin
$ git add foo.bin
$ git commit -m init
$ git lfs checkout
Checking out LFS objects: 100% (1/1), 0 B | 0 B/s, done. 
```

I believe this stems from work in PR #4525, and PR #4681 adjusted `git lfs ls-files` to filter out empty files, which is why you don't see them there.

**Comment by gpakosz** [comment]
> Ah, that's a good point, I forgot that since v3.x Git LFS recognizes zero-byte files regardless of their tracked status.

I'm not sure I understand why it's the case.
What I know is that in this repository I'm working with, which I unfortunately cannot share, `git lfs prune` takes a lot of time, to do nothing in the end. It even looks like it's hanging as it outputs nothing. This is what got me curious in the first place.

**Reply by chrisd8088** [reply]
I don't think empty files would have any impact on what `git lfs prune` is doing.  You might want to turn on `GIT_TRACE=1` to see if that provides any indication of where `git lfs prune` is working or stalling.

As for empty files, the specific change in v3.0 comes from git-lfs/git-lfs@efaf9a2d42636e58373351c40402d9d2a08fd1eb, I believe.

**Comment by gpakosz** [comment]
`git lfs prune` appears to be stalling because the following takes a lot of time. Note that it completes eventually.
```
$ GIT_TRACE=1 git '-c' 'filter.lfs.smudge=' '-c' 'filter.lfs.clean=' '-c' 'filter.lfs.process=' '-c' 'filter.lfs.required=false' 'log' '--branches' '--tags' '--not' '--remotes=origin' '--no-ext-diff' '--no-textconv' '--color=never' '-G' 'oid sha256:' '-p' '-U12' '--format=lfs-commit-sha: %H %P'
```

Running `watch lsof -a -p <pid of git process mentioned above>` shows that various `.idx` and `.pack` files are being processed.

It looks like it's inspecting revisions to see if it's safe to prune these empty files.

**Reply by chrisd8088** [reply]
Hmm.  This boils down to a `git log -G 'oid sha256:'` command, I think.

The other arguments unset the Git LFS configuration, adjust output formatting, and control the specific set of revisions being examined, especially the `--branches --tags --not --remotes=origin` part, which means, as [described](https://git-scm.com/docs/git-log#Documentation/git-log.txt-codegitlog--branches--not--remotesorigincode) by the `git-log(1)` manual page, "all commits that are in any of local branches [and local tags] but not in any of remote-tracking branches for `origin` (what you have that `origin` doesn't)".

So if you have a lot of local branches and/or tags, that might examine a lot of commits.  But it should only be looking for commits where the patch text matches the string `oid sha256:`; that's the `-G` option.  In other words, it's looking for commits where the blob contents added or changed or removed a Git LFS pointer file, since those always include the string `oid sha256:`.  (Of course, other blobs may happen to have that string too; false positives get identified later and ignored.)

Since empty blobs can't have that string, I don't think it's possible they are the source of your slow experience.  I suspect it's more likely there is just a lot of Git history in your local branches and/or tags, and it all has to be examined when you prune.

**Comment by gpakosz** [comment]
> Since empty blobs can't have that string, I don't think it's possible they are the source of your slow experience. I suspect it's more likely there is just a lot of Git history in your local branches and/or tags, and it all has to be examined when you prune.

But there's nothing to prune to begin with, right? 🙃

**Reply by chrisd8088** [reply]
I'm not quite sure I understand the context here.  If you aren't using Git LFS at all, I would just not run `git lfs prune`.

But if you do run the command, it looks through your Git history.  It's not a matter of whether there are empty files or not.  One can see this by doing:
```
$ git init test
$ cd test
$ echo foo >foo.txt
$ git add foo.txt
$ git commit -m init

$ GIT_TRACE=1 git lfs prune 2>&1 | grep oid
14:16:46.335633 trace git-lfs: exec: git '-c' 'filter.lfs.smudge=' '-c' 'filter.lfs.clean=' '-c' 'filter.lfs.process=' '-c' 'filter.lfs.required=false' 'log' '--branches' '--tags' '--not' '--remotes=origin' '--no-ext-diff' '--no-textconv' '--color=never' '-G' 'oid sha256:' '-p' '-U12' '--format=lfs-commit-sha: %H %P'
```

Arguably that could be improved if there's nothing in `.git/lfs/objects`; by all means open an issue if you think it's valuable.  This behaviour is not related to empty files, though.

**Reply by chrisd8088** [reply]
Patches are also always welcome!

**Comment by gpakosz** [comment]
> I would just not run `git lfs prune`

True.

> This behaviour is not related to empty files, though

I indeed misread your explanation about `git log -G 'oid sha256:'`.

> Arguably that could be improved if there's nothing in `.git/lfs/objects`

What's the most robust way of achieving "invoke `git lfs prune` only if the repository uses LFS"? Inspecting the output of `git ls-files ':(attr:filter=lfs)'` or inspecting `.git/lfs/objects`?

> Patches are also always welcome!

Noted.

**Reply by chrisd8088** [reply]
> What's the most robust way of achieving "invoke `git lfs prune` only if the repository uses LFS"? Inspecting the output of `git ls-files ':(attr:filter=lfs)'` or inspecting `.git/lfs/objects`?

Good question ... I suppose it depends a bit on your use case, but in general, I would think you might want to look at `.git/lfs/objects`.  The `git lfs prune` command is really intended to control the number of locally-cached Git LFS objects, i.e., the ones in `.git/lfs/objects`.

Since a repository's history may include Git LFS objects and corresponding entries in `.gitattributes` which are not present in the currently checked-out working tree, just looking at a particular revision of `.gitattributes` may not help you determine if Git LFS has been used in the repository at all.  (Indeed, commands like `git lfs migrate export` will go through your history and search for  `filter=lfs` entries in `.gitattributes` at any point in history.)

But in an arbitrary repository that's being actively used for development, with various branches checked in and out, copies of Git LFS objects will accumulate in `.git/lfs/objects`, only some of which are in the working tree at any given time.  Also, `.git/lfs/objects` will hold copies of any Git LFS objects which have been locally committed but not yet pushed to a remote, and these may not all be in the current working tree either.

So over time, developers may want to trim down the amount of disk space consumed by these cached copies.  That's what `git lfs prune` is for; it lets you say, for instance, "get rid of all cached objects that have been pushed to a remote and which are not the in working tree and which are more than 3 days old", etc.  (The manual page describes the exact variety of options and rules.)

Hence, if you're trying to decide whether to run `git lfs prune`, I would check for the existence of files in `.git/lfs/objects`, because that's the only location from which it will prune files, if it can.

**Reply by chrisd8088** [reply]
To put it another way: if you don't have `.git/lfs/objects` or it's empty, then there's no value in running `git lfs prune`; there's nothing it can prune.

**Reply by gpakosz** [reply]
Thanks for the detailed answers @chrisd8088. Much appreciated.

**Comment by cov393** [comment]
When `git lfs ls-files` returns no output but `git lfs checkout` reports that it is checking out LFS objects, it means that your repository is using Git LFS to track large files, but the files themselves are not currently checked out on your local machine.

The `git lfs ls-files` command lists all the files that are tracked by Git LFS and their associated pointer files, whereas `git lfs checkout` retrieves and checks out the actual large files from Git LFS. If you have just cloned the repository or switched to a branch that contains LFS files, you will need to run `git lfs checkout` to download and checkout the LFS files.

In your case, it seems that Git LFS is correctly tracking the large files in your repository, and you just need to run `git lfs checkout` to retrieve and checkout the actual files. The output you provided indicates that there are 15 LFS objects that need to be checked out.

Once you run `git lfs checkout`, Git LFS will retrieve the LFS objects and replace the pointers with the actual large files on your local machine. After that, you should be able to see the actual files in your working directory.
